report/renderer/util.js

```diff
@@ -230,6 +233,79 @@ class Util {
     return clone;
   }
 
+  /**
+   * Given an audit's details, identify and return a URL locator function that
+   * can be called later with an `item` to extract the URL of it.
+   * @param {LH.FormattedIcu<LH.Audit.Details.TableColumnHeading[]>} headings
+   * @return {{(item: LH.FormattedIcu<LH.Audit.Details.TableItem>): string|undefined}=}
+   */
+  static getUrlLocatorFn(headings) {
+    // The most common type, valueType=url.
+    const urlKey = headings.find(heading => heading.valueType === 'url')?.key;
+    if (urlKey) {
+      // Return a function that extracts item.url.
+      return (item) => item[urlKey]?.toString();
+    }
+
+    // The second common type, valueType=source-location.
+    const srcLocationKey = headings.find(heading => heading.valueType === 'source-location')?.key;
+    if (srcLocationKey) {
+      // Return a function that extracts item.source.url.
+      return (item) => {
+        const sourceLocation = item[srcLocationKey];
+        if (typeof sourceLocation === 'object' && sourceLocation.type === 'source-location') {
+          return sourceLocation.url;
+        }
+      };
+    }
+
+    // More specific tests go here, as we need to identify URLs in more audits.
+  }
+
+  /**
+   * Mark TableItems/OpportunityItems with entity names.
+   * @param {LH.Result.EntityClassification|undefined} entityClassification
+   * @param {import('../../types/lhr/audit-result').Result} audit
```

Comment:
@brendankenny @adamraine is it ok to make this change so we can use this type more directly?

```
diff --git a/report/types/html-renderer.d.ts b/report/types/html-renderer.d.ts
index f7665fd2d..f5925e0c8 100644
--- a/report/types/html-renderer.d.ts
+++ b/report/types/html-renderer.d.ts
@@ -5,6 +5,7 @@
  */
 
 import AuditDetails from '../../types/lhr/audit-details';
+import {Result as AuditResult} from '../../types/lhr/audit-result.js';
 import {FormattedIcu as FormattedIcu_, IcuMessage as IcuMessage_} from '../../types/lhr/i18n';
 import LHResult from '../../types/lhr/lhr';
 import ReportResult_ from './report-result';
@@ -24,6 +25,7 @@ declare global {
 
     module Audit {
       export import Details = AuditDetails;
+      export type Result = AuditResult;
     }
 
     export import Treemap = Treemap_;
```

Reply:
`LH.Audit.Result` (or typedef it if you don't want it inline)

Reply:
that doesn't work without my patch
<img width="381" alt="image" src="https://user-images.githubusercontent.com/4071474/213305706-50de8602-2f40-421b-b599-0d4181fe25bc.png">
<img width="440" alt="image" src="https://user-images.githubusercontent.com/4071474/213305744-6c3d89af-4082-49e9-b507-63642a6b2b24.png">

Reply:
Oh, whoops, in the report, right. Importing like that or a typedef is fine IMO, but that will work.